File: Interview_Prep_App/native_java/javatemplate/src/main/java/com/tc2r/javatemplate/ui/main/MainActivity.java
package com.tc2r.javatemplate.ui.main;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentTransaction;

import com.tc2r.javatemplate.R;
import com.tc2r.javatemplate.repositories.Tc2rGithubRepository;
import com.tc2r.javatemplate.repositories.models.Answer;
import com.tc2r.javatemplate.repositories.models.Question;
import com.tc2r.javatemplate.ui.questions.DialogFragment;
import com.tc2r.javatemplate.ui.questions.QuestionFragment;
import com.tc2r.javatemplate.ui.score.ScoreActivity;
import com.tc2r.javatemplate.util.OnFragmentInteractionListener;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Random;
import java.util.concurrent.Callable;


public class MainActivity extends AppCompatActivity implements OnFragmentInteractionListener {
    // Static Variables
    private final static int QUIZ_SIZE = 5;

    // UI Variables
    private TextView titleTv, scoreTv;


    // Variables
    private Tc2rGithubRepository tc2rGithubRepository;
    private ArrayList<Question> quizList;
    private ArrayList<Question> testList;
    private ArrayList<Answer> answersList;
    private Random random;
    private int currentQuestion = 0;
    private int numOfCorrect = 0;
    private int scorePer = 0;
    private double pointPerQ, score;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tc2rGithubRepository = new Tc2rGithubRepository();

        // Initialize and assignments
        titleTv = findViewById(R.id.title_tv);
        scoreTv = findViewById(R.id.score_tv);
        testList = new ArrayList<>();
        quizList = new ArrayList<>();
        answersList = new ArrayList<>();
        random = new Random();

        // FIXME: Call the below code off of the main thread in order not to freeze the main thread
        //  while information is being retrieved.
        getAnswers();
        getQuestions();
        createQuiz();
    }

    public void getAnswers() {
        answersList.addAll(tc2rGithubRepository.getAnswersList());
    }

    public void getQuestions() {
        quizList.addAll(tc2rGithubRepository.getQuestions());
    }

    private void createQuiz() {
        Log.i(" Size: ", "QuestionList is: " + quizList.size());
        Log.i(" Size: ", "AnswerList is: " + answersList.size());

        // set booleanArray to be same size as quizList
        boolean[] selectedQuestion = new boolean[quizList.size()];

        // set an int to a random number in the quizList
        int randNum = random.nextInt(quizList.size());
        int i = 0;

        // sets the score system for quiz.
        // TODO: 5/31/2017 maybe remove this and simply divide correct answers by total questions
        pointPerQ = 1.0 / QUIZ_SIZE;

        // prevent from having repeated questions in a single quiz
        while(i < QUIZ_SIZE) {
            // if boolean at randNum in selectedQuestion is false
            if(!selectedQuestion[randNum] && quizList.get(randNum).getQuestionType().equals("multi")) {

                // Add position randNum to test list;
                testList.add(quizList.get(randNum));
                // set this question selected to true.
                selectedQuestion[randNum] = true;
                i++;
            } else {
                // if question already selected, change randNum;
                randNum = random.nextInt(quizList.size());
            }
        }

        // On first run, start quiz without updating score
        nextQuestion(false);
    }

    public String goToNextQuestion(boolean isAnswerCorrect) {
        nextQuestion(isAnswerCorrect);
        return "True";
    }

    public void showDetails(Answer answer, boolean isAnswerCorrect) {
        // Show a details dialog fragment on top of the current screen.
        DialogFragment myDialogFragment = new DialogFragment(answer, new Callable<String>() {
            public String call() {
                // after user dismisses the dialogfragment, move on to next question.
                return goToNextQuestion(isAnswerCorrect);
            }
        });
        myDialogFragment.show(getSupportFragmentManager(), "details");
    }


    public void nextQuestion(boolean correctAnswer) {
        // if previous question was answered correctly
        // update variables accordingly.
        if(correctAnswer) {
            numOfCorrect++;
            score += pointPerQ;
            scorePer = (int) (score * 100);
            scoreTv.setText(String.format(Locale.US, "%s%d", getString(com.tc2r.sharedresources.R.string.score_display_text), scorePer));
        }

        // if quiz is not complete, continue quiz with new QuestionFragment
        if(currentQuestion < QUIZ_SIZE) {
            // Fragments and Model Variables
            QuestionFragment newFragment = QuestionFragment.newInstance(testList.get(currentQuestion), answersList);
            currentQuestion++;
            titleTv.setText(String.format(Locale.US, "%s%d of %d", getString(com.tc2r.sharedresources.R.string.question_display_text), currentQuestion, QUIZ_SIZE));
            // UI Variables
            LinearLayout fragContainer = findViewById(R.id.fragment_container);
            FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
            ft.replace(fragContainer.getId(), newFragment);
            ft.commit();
        } else {
            // Quiz is over, go to final page!
            // create intent
            Intent intent = new Intent(this, ScoreActivity.class);

            // add variables to send.
            intent.putExtra("scorePercentage", scorePer);
            intent.putExtra("quizSize", QUIZ_SIZE);
            intent.putExtra("numCorrect", numOfCorrect);

            // use intent.
            startActivity(intent);
        }
    }

    @Override
    public void fragmentInitialized() {

    }
}

File: Interview_Prep_App/native_kotlin/retrofit2_moshi_Androidlion48/src/main/java/com/interviewprep/moshi_androidlion48/ui/main/MainActivity.kt
package com.interviewprep.moshi_androidlion48.ui.main

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.interviewprep.kotlinretrofit.util.OnFragmentInteractionListener
import com.interviewprep.moshi_androidlion48.R
import com.interviewprep.moshi_androidlion48.repository.models.Answer
import com.interviewprep.moshi_androidlion48.repository.models.Question
import com.interviewprep.moshi_androidlion48.repository.tc2rgithubrepository.Tc2rGithubRepository
import com.interviewprep.moshi_androidlion48.ui.questions.DialogFragment
import com.interviewprep.moshi_androidlion48.ui.questions.QuestionFragment
import com.interviewprep.moshi_androidlion48.ui.score.ScoreActivity
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.*

// fixme: (Optional) Use Databinding, Dependency Injection, or ViewBinding to refactor this fragment.
class MainActivity : AppCompatActivity(), OnFragmentInteractionListener {
    private val TAG = "MAINTEST"

    companion object {
        // Static Variables
        private const val QUIZ_SIZE = 5
    }

    private val tc2rGithubRepository: Tc2rGithubRepository by lazy {
        Tc2rGithubRepository()
    }

    // UI Variables
    private var fragContainer: LinearLayout? = null
    private var titleTv: TextView? = null
    private var scoreTv: TextView? = null

    // Fragments and Model Variables
    private var newFragment: QuestionFragment? = null
    private var quizList: ArrayList<Question>? = null
    private var testList: ArrayList<Question>? = null
    private var answersList: ArrayList<Answer>? = null
    private lateinit var selectedQuestion: BooleanArray

    // Variables
    private var random: Random? = null
    private var currentQuestion = 0
    private var numOfCorrect = 0
    private var scorePer = 0
    private var pointPerQ = 0.0
    private var score = 0.0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Initialize and assignments
        titleTv = findViewById(R.id.title_tv)
        scoreTv = findViewById(R.id.score_tv)
        testList = ArrayList()
        quizList = ArrayList()
        answersList = ArrayList()
        random = Random()

        // FIXME: Call the below code off of the main thread in order not to freeze the main thread
        //  while information is being retrieved.
        val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
            Log.e("Tc2r1", throwable.message.toString())
        }

        lifecycleScope.launch(coroutineExceptionHandler) {
            answersList = ArrayList(tc2rGithubRepository.getAndroidAnswers().body()?.answers ?: emptyList())
            quizList = ArrayList(tc2rGithubRepository.getAndroidQuestions().body()?.questions ?: emptyList())

            launch(Dispatchers.Main) {
                createQuiz()
            }
        }
    }

    private fun createQuiz() {
        Log.i(TAG, "QuestionList is: " + quizList!!.size)
        Log.i(TAG, "AnswerList is: " + answersList!!.size)
        // set booleanArray to be same size as quizList
        selectedQuestion = BooleanArray(quizList!!.size)

        // set an int to a random number in the quizList
        var randNum = random!!.nextInt(quizList!!.size)
        var i = 0

        // sets the score system for quiz.
        // TODO: 5/31/2017 maybe remove this and simply divide correct answers by total questions
        pointPerQ = 1.0 / QUIZ_SIZE
        while (i < QUIZ_SIZE) {
            // Log.wtf("testList Size: ", String.valueOf(testList.size()) + " Vs "+ String.valueOf(quizSize) );
            // if boolean at randNum in selectedQuestion is false
            if (!selectedQuestion[randNum]) {
                // Add position randNum to test list;
                testList!!.add(quizList!![randNum])
                // set this question selected to true.
                selectedQuestion[randNum] = true
                i++
            } else {
                // if question already selected, change randNum;
                randNum = random!!.nextInt(quizList!!.size)
            }
        }
        // On first run, start quiz without updating score
        nextQuestion(false)
    }

    fun showDetails(answer: Answer, isAnswerCorrect: Boolean) {
        // Show a details dialog fragment on top of the current screen.
        val dialogFragment = DialogFragment.newInstance(
            answer = answer,
            func = {
                // after user dismisses the dialogFragment, move on to the next question.
                nextQuestion(isAnswerCorrect)
            }
        )
        dialogFragment.show(supportFragmentManager, "details")
    }

    private fun nextQuestion(correctAnswer: Boolean) {
        // if previous question was answered correctly
        // update variables accordingly.
        if (correctAnswer) {
            numOfCorrect++
            score += pointPerQ
            scorePer = (score * 100).toInt()
            scoreTv!!.text = getString(com.tc2r.sharedresources.R.string.score_display_text) + scorePer.toString()
        }
        // if quiz is not complete, continue quiz with new QuestionFragment
        if (currentQuestion < QUIZ_SIZE) {
            newFragment = QuestionFragment.newInstance(testList!![currentQuestion], answersList!!)
            currentQuestion++
            titleTv!!.text = getString(com.tc2r.sharedresources.R.string.question_display_text) + Integer.toString(currentQuestion) + " of " + Integer.toString(QUIZ_SIZE)
            fragContainer = findViewById(R.id.fragment_container)
            val ft = supportFragmentManager.beginTransaction()
            ft.replace(fragContainer!!.id, newFragment!!)
            ft.commit()
        } else {
            // Quiz is over, go to final page!
            // create intent
            val intent = Intent(this, ScoreActivity::class.java)

            // add variables to send.
            intent.putExtra("scorePercentage", scorePer)
            intent.putExtra("quizSize", QUIZ_SIZE)
            intent.putExtra("numCorrect", numOfCorrect)

            // use intent.
            startActivity(intent)
        }
    }

    override fun fragmentInitialized() {}
}
